tools/generate_validation_tolerances.py

```diff
@@ -0,0 +1,64 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+This script reproduces the computation used for generating the validation tolerances in csrc/validator_utils.h:
+https://github.com/NVIDIA/Fuser/blob/892b7ac5646b3b873f13f2b2ea1db15fa9b9effb/csrc/validator_utils.h#L25-L46
+
+The computation (randn + mul + sum) is repeated a large number of times.
+The base tolerance is the max error seen when comparing the desired precision type with the next higher precision.
+The final tolerances are computed by adding a safety factor of 3/4 to this base tolerance.
+
+The output of this file is a .npy file corresponding to each dtype, containing a {size: base_tolerances} dict.
+
+Note: The script may take a long time for a high number of num_iters, so it may be useful to only run the script for one dtype at a time. Modify `dtype_to_ref_dtypes` accordingly.
+
+To run: python tools/generate_validation_tolerances.py
+To load the files:
+    ```
+    import numpy as np
+    base_tol_dict = np.load(file_name.npy, allow_pickle=True).item()
+    ```
+"""
+
+import numpy as np
+import torch
+from datetime import datetime
+
+sizes = [2**i for i in range(2, 22)]  # {4, 2097152}
```

Comment:
Question: IIRC, the reduction size is computed with respect to fusion inputs, so it tends to grow very fast. For a two-layer MLP (https://github.com/NVIDIA/Fuser/pull/2905), the reduction size of the second linear output is already `4 * hidden_size * hidden_size`, close to 2M. I'd imagine the whole transformer block will generate an even larger reduction size, much larger than the max size specified here. What's the implication of getTolerance querying a size larger than the max?

Reply:
It would use twice the max error in the list(https://github.com/NVIDIA/Fuser/blob/1158543c9bd16190fe5820b41a7f4c7e891bc5d9/csrc/validator_utils.cpp#L134-L138). So far, for the unit tests, these existing sizes have been sufficient.
If we have very few examples with larger (~2M) reduction sizes, it would be simpler to set a threshold manually. If we have several such examples, we may benefit from having more cases.

Reply:
> If we have several such examples, we may benefit from having more cases.

In case you are looking for concrete use cases, I believe https://github.com/NVIDIA/Fuser/blob/main/tests/cpp/test_multidevice_transformer.cpp#L569-L578 will give you reduction sizes much larger than 2M, when you change validation to use testValidate.

Reply:
The other issue with MLP is we have compounding error from the running consecutive ops